File: Python/FloodFill.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        original_color = image[sr][sc]
        if original_color == color:
            return image
        rows = len(image)
        cols = len(image[0])
        def dfs(r, c):
            #Check boundaries
            if r < 0 or r >= rows or c < 0 or c >= cols:
                return
            #Color check is this pixel already painted? 
            #Also check if the starting color have the same color of its adjacent or simply detects a wall
            if image[r][c] != original_color:
                return
            #Paint the pixel
            image[r][c] = color
            #Spread the paint
            dfs(r-1, c) #Up
            dfs(r+1,c)  #Down
            dfs(r, c-1) #Left
            dfs(r, c+1) #Right
        # Start the recursion from the pixel we originally clicked
        dfs(sr, sc)
        return image

File: CPP/FloodFill.cpp
class Solution {
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        int original_color = image[sr][sc];
        if(original_color != color){
            dfs(image, sr, sc, original_color, color);
        }
        return image;
    }
    private:
        void dfs(vector<vector<int>>& image, int r, int c, int original_color, int new_color ){
            //Boundary check
            if(r < 0 || r >= image.size() || c < 0 || c >= image[0].size()){
                return;
            }
            //Color check and wall check
            if(image[r][c] != original_color){
                return;
            }
            image[r][c] = new_color;
            dfs(image, r-1, c, original_color, new_color); //UP
            dfs(image, r+1, c, original_color, new_color); //DOWN
            dfs(image, r, c-1, original_color, new_color); //LEFT
            dfs(image, r, c+1, original_color, new_color); //RIGHT
        }
};